docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -0,0 +1,590 @@
+
+Use cases for a verifier plugin to interact with the OE SDK framework:
+- Register a verifier plugin.
+  - Helper function
+  `oe_result_t oe_register_verifier(oe_verifier_t* plugin, const void* config_data, size_t config_data_size)`.
+  - The same definition as in the OE SDK V0.9 release, except that this function
+  is not part of the public API.
+- Unregister a verifier plugin.
+  - Helper function
+  `oe_result_t oe_unregister_verifier(oe_verifier_t* plugin)`.
+  - The same definition as in the OE SDK V0.9 release, except that this function
+  is not part of the public API.
+
+In the OE SDK V0.10 release, initialization of verifier plugin libraries is
+triggered with the application call to `oe_verifier_initialize()`.
+The initialization function of every linked library enumerates its supported
+evidence formats and registers their plugins with the OE SDK framework via
+helper function `oe_register_verifier()`.
+
+# Implementation of SGX Plugins
+
+## Existing OE SDK V0.9 Implementation
+
+Only one attester plugin and one verifier plugin are implemented,
+for a single SGX ECDSA evidence format identified by `OE_SGX_PLUGIN_UUID`.
+
+The attester plugin entry points are implemented in
+[enclave/sgx/attester.c](https://github.com/openenclave/openenclave/blob/v0.9.x/enclave/sgx/attester.c).
+The plugin entry points are returned in an `oe_attester_t` structure by the
+implementation of the public API function `oe_sgx_plugin_attester()`. This
+attester plugin implementation is linked to the OE SDK enclave-side static
+library.
+
+The verifier plugin entry points are implemented in
+[common/sgx/verifier.c](https://github.com/openenclave/openenclave/blob/v0.9.x/common/sgx/verifier.c).
+The plugin entry points are returned in an `oe_verifier_t` structure by the
+implementation of the public API function `oe_sgx_plugin_verifier()`. This
+verifier plugin implementation is linked to both the OE SDK enclave-side and host-side
+static libraries.
+
+## Proposed V0.10 Implementation
+
+The OE SDK V0.10 release will have enclave-side SGX attester plugins for
+generation of SGX evidence in local, ECDSA, and EPID formats. These formats
+are defined as:
+- `OE_SGX_LOCAL_ATTESTATION_PLUGIN_UUID`: for SGX local attestation.
+- `OE_SGX_ECDSA_P256_PLUGIN_UUID`: for SGX ECDSA-p256 evidence format
+  - Note: this is the same as existing `OE_SGX_PLUGIN_UUID`.
+- `OE_SGX_EPID_LINKABLE_PLUGIN_UUID`: for SGX linkable EPID evidence format.
+- `OE_SGX_EPID_UNLINKABLE_PLUGIN_UUID`: for SGX unlinkable EPID evidence format.
+
+These plugins will be implemented by a single set of plugin libraries,
+composed of:
+- One enclave-side SGX evidence generation plugin library.
+  - It implements the OE SDK public API function
+  `oe_attester_initialize()` for attester plugins enumeration and
+  registration, and `oe_attester_shutdown()`.
+- One host-side SGX evidence generation plugin library.
+  - It implements OCALLs for the enclave-side library to access
```

Comment:
Is doing OCALLs safe to do?  Will make this less secure given that the enclave will rely on the host to get the evidence?  This also adds overhead.

Reply:
Resolving this given that there is already existing ocalls.  But I highly suggest avoiding ocalls given the added overhead and increased attack footprint.